models/models_methods.go

```diff
@@ -450,6 +470,74 @@ func (config *BitriseDataModel) Validate() ([]string, error) {
 	}
 	// ---
 
+	// pipelines
+	for ID, pipeline := range config.Pipelines {
+		if ID == "" {
+			return warnings, fmt.Errorf("invalid pipeline ID (%s): empty", ID)
+		}
+
+		r := regexp.MustCompile(`[A-Za-z0-9-_.]+`)
+		if find := r.FindString(ID); find != ID {
+			warnings = append(warnings, fmt.Sprintf("invalid pipeline ID (%s): doesn't conform to: [A-Za-z0-9-_.]", ID))
+		}
+
+		if len(pipeline.Stages) == 0 {
+			return warnings, fmt.Errorf("pipeline (%s) should have at least 1 stage", ID)
+		}
+
+		for _, pipelineStage := range pipeline.Stages {
+			pipelineStageID, _, err := GetStageIDStageDataPair(pipelineStage)
+			if err != nil {
+				return warnings, err
+			}
+			found := false
+			for stageID := range config.Stages {
+				if stageID == pipelineStageID {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return warnings, fmt.Errorf("stage (%s) defined in pipeline (%s), but does not exist", pipelineStageID, ID)
+			}
+		}
+	}
+	// ---
+
+	// stages
+	for ID, stage := range config.Stages {
+		if ID == "" {
```

Comment:
Please extract ID validation (empty check and regexp conformance) to a separate functions since validation is the same for `pipelineID`, `stageID` and `workflowID`

Reply:
I think you have forgotten about the ID validation extraction.

Reply:
Nope, it's [here](https://github.com/bitrise-io/bitrise/pull/735/files#diff-1bb32c53b12166a3a66db70cbe1be4356cbc8adac0b84d14887022d830f8594aR597-R607)